source/Ship.cpp

```diff
@@ -3511,46 +3508,55 @@ const vector<weak_ptr<Ship>> &Ship::GetEscorts() const
 // cues and try to stay with it when it lands or goes into hyperspace.
 void Ship::AddEscort(Ship &ship)
 {
-	// Cache the maximum speeds of escorts so that the parent can stay
-	// below this speed when the parent wants to keep all escorts together.
-	if(escorts.empty() || escortsVelocity < 0.)
-		escortsVelocity = ship.MaxVelocity();
-	else
-		escortsVelocity = min(escortsVelocity, ship.MaxVelocity());
-	
+	RegisterEscort(ship);
 	escorts.push_back(ship.shared_from_this());
 }
 
 
 
 void Ship::RemoveEscort(const Ship &ship)
 {
+	// Reset this value, we will re-scan the list of escorts to set the
+	// escorts velocity based on remaining active escorts.
+	escortsVelocity = -1.;
+
 	auto it = escorts.begin();
-	for( ; it != escorts.end(); ++it)
-		if(it->lock().get() == &ship)
-		{
-			escorts.erase(it);
-			break;
-		}
-	
-	// If we remove the slowest escort from a fleet, then the fleet might
-	// be able to speed up.
-	if(ship.MaxVelocity() == escortsVelocity)
+	while(it != escorts.end())
 	{
-		escortsVelocity = -1.;
-
-		it = escorts.begin();
-		for( ; it != escorts.end(); ++it)
-		{
-			auto escort = it->lock();
-			if(escort)
-			{
-				if(escortsVelocity < 0.)
-					escortsVelocity = escort->MaxVelocity();
-				else
-					escortsVelocity = min(escortsVelocity, escort->MaxVelocity());
+		auto escort = it->lock().get();
+		if(escort){
+			if(escort == &ship){
+				escorts.erase(it);
+				continue;
 			}
+			else
+				// Re-scan existing escorts to cache data like the
+				// escorts velocity.
+				RegisterEscort(ship);
 		}
+		++it;
+	}
+}
+
+
+
+// Store relevant cached data for the given escort.
+void Ship::RegisterEscort(const Ship &ship)
+{
+	// Cache the maximum speed that the parent should stay below to keep
+	// all escorts together.
+	// We don't cache the speeds of escorts that have 0 velocity to avoid
+	// having the parent get stuck when one of the escorts doesn't have
+	// regular thrust.
+	// We also don't cache the speeds of carried ships, since they are often
+	// docked and the carrier waits for them during docking already.
+	double eV = ship.MaxVelocity() * 0.9;
+	if(!ship.CanBeCarried() && eV > 0.)
```

Comment:
move the definition of `eV` inside the if-block.

Note that you could also account for carried escort state in a similar manner to how we handle `carriedMass` -- update on launch & carry.

Reply:
Moved the definition of eV.

I considered updating on launch & carry. For deploy this would be fine; if the deployed ship is slower than the current lowest escort-speed, then update the esocrt-speed to the lower value. For docking it could be a bit trickier; if you have a large number of slow carried escorts, then each docking ship could trigger a loop to determine the next slowest escort.